### ISSUE TITLE
Documentation for BinaryThresholdPredictor

### ISSUE THREAD

**Issue by tpoisot** [alan-turing-institute/MLJ.jl#973]
**Is your feature request related to a problem? Please describe.**
I'm trying to use a `BinaryThresholdPredictor` to do thresholded classification, and I am not sure from the documentation how it should actually be done. For now, I am using an `EvoTreeGaussian` which returns a distribution (a `Normal`), but I can't get the `BinaryThresholdPredictor` to wrap it. The documentation just says it's possible, but is unclear.

**Describe the solution you'd like**
Update https://alan-turing-institute.github.io/MLJ.jl/dev/more_on_probabilistic_predictors/#MLJModels.BinaryThresholdPredictor to have an actual example of use, and a better way to identify models that can be wrapped this way?

**Describe alternatives you've considered**
I've pretty much tried all of the combinations of different models, wrapping the model, the machine, the type, etc...

**Additional context**
N/A

**Comment by tpoisot** [comment]
One additional thing I'd be curious about is to see a use-case of tuning the threshold for a `BinaryThresholdPredictor` using `curve`; I managed to get an example going (wrapping a pipeline that was essentially PCA -> BRT classifier, so at least I have a better understanding of what a distribution at the end should look like), but I haven't found the secret incantation to fine-tune the threshold. Is there a way?

**Comment by ablaom** [comment]
@tpoisot Kind of you to report, thanks. 

`BinaryThresholdPredictor` is for wrapping deterministic (point) predictors whose target scitype is `AbstractArray{<:Multiclass{2}}` - essentially deterministic classifiers and some outlier detection models. It takes a deterministic model and makes it probabilistic. I agree the docstring is lacking, and have tagged the issue "docs"

Sounds like you're trying to go in the other direction, probabilistic -> deterministic *and* you want a binary prediction (classification) but are starting with a regressor? Are you wanting to threshold on the mean prediction? (Note you can get the mean directly with `predict_mean` instead `predict`.)

This may be related: https://github.com/JuliaAI/MLJBase.jl/issues/760

Assuming I understand your use-case correctly, one could build a [learning network](https://alan-turing-institute.github.io/MLJ.jl/dev/composing_models/#Learning-Networks) to do what you want, and make the threshold part of the exported composite model hyperparameters:

```julia
using MLJ
import MLJBase

mutable struct CompositeD <: DeterministicComposite
    regressor # that makes probabilistic predictions
    threshold
end

#Here's a helper to do the actual thresholding:
classes = coerce(["low", "high"], OrderedFactor{2})
split_on(threshold, y) = map(y) do x
    x < threshold ? first(classes) : last(classes)
end

julia> split_on(0.5, [0.2, 0.3, 0.6])
3-element CategoricalArrays.CategoricalArray{String,1,UInt32}:
 "low"
 "low"
 "high"

function MLJBase.fit(composite::CompositeD, verbosity, X, y)

    Xs = source(X)
    ys = source(y)

    mach = machine(composite.regressor, Xs, ys)
    y_continuous = predict_mean(mach, Xs)

    # we use @node here because `y_continuous` is a node:
    y_binary = @node split_on(composite.threshold, y_continuous)

    network_machine = machine(Deterministic(), Xs, ys, predict=y_binary)
    return!(network_machine, composite, verbosity)
end

# Demonstration on Boston dataset:

X, y = @load_boston # a table and a vector

rgs = (@load EvoTreeGaussian)()
μ = mean(y)
composite_d = CompositeD(rgs, μ)

# check we get disrete preditions:
mach = machine(composite_d, X, y) |> fit!

julia> predict(mach, rows=1:3)
3-element CategoricalArrays.CategoricalArray{String,1,UInt32}:
 "high"
 "low"
 "high"
```

Now this wrapped model will  be non-standard as far as optimization is concerned, because the training target is continuous, but the predictions are discrete. If you wanted to wrap again using `TunedModel` to optimize `threshold` then you will need to specify a custom metric `m(yhat, y)` to articulate what you want to do, right? (I didn't quite understand your "tuning the threshold for a BinaryThresholdPredictor using curve".  Maybe you want to elaborate on that.)

**Comment by tpoisot** [comment]
Thanks for the super detailed answer, @ablaom -- I think as a starting point, I'd like to be able to reproduce *e.g.* https://machinelearningmastery.com/threshold-moving-for-imbalanced-classification/ using MLJ. This example uses logistic regression, then optimizes the threshold to maximize Youden's J. Maybe the `BinaryThresholdPredictor` isn't the right approach?

**Comment by tpoisot** [comment]
Amazing! I experimented with `BinaryThesholdPredictor` some more this afternoon, and it's all good now. Your latest example helps *a lot* too.

**Comment by ablaom** [comment]
Okay, then we can add some variant of that example to the docstring.
